yarn-project/p2p/src/services/tx_collection/fast_tx_collection.ts

```diff
@@ -0,0 +1,329 @@
+      const logCtx = {
+        ...blockInfo,
+        errorMessage: err instanceof Error ? err.message : undefined,
+        missingTxs: [...request.missingTxHashes].map(txHash => txHash.toString()),
+      };
+      if (err instanceof Error && err.name === 'TimeoutError') {
+        this.log.warn(`Timed out collecting txs for ${request.type} at slot ${blockInfo.slotNumber}`, logCtx);
+      } else if (err instanceof Error && err.name === 'AbortError') {
+        this.log.warn(`Aborted collecting txs for ${request.type} at slot ${blockInfo.slotNumber}`, logCtx);
+      } else {
+        this.log.error(`Error collecting txs for ${request.type} for slot ${blockInfo.slotNumber}`, err, logCtx);
+      }
+    } finally {
+      // Ensure no unresolved promises and remove the request from the set
+      request.promise.resolve();
+      this.requests.delete(request);
+    }
+  }
+
+  /**
+   * Starts collecting txs from all configured nodes. We send `txCollectionFastMaxParallelRequestsPerNode` requests
+   * in parallel to each node. We keep track of the number of attempts made to collect each tx, so we can prioritize
+   * the txs that have been requested less often whenever we need to send a new batch of requests. We ensure that no
+   * tx is requested more than once at the same time to the same node.
+   */
+  private async collectFastFromNodes(request: FastCollectionRequest, opts: { deadline: Date }): Promise<void> {
+    if (this.nodes.length === 0) {
+      return;
+    }
+
+    // Keep a shared priority queue of all txs pending to be requested, sorted by the number of attempts made to collect them.
+    const attemptsPerTx = [...request.missingTxHashes].map(txHash => ({ txHash, attempts: 0, found: false }));
+
+    // Returns once we have finished all node loops. Each loop finishes when the deadline is hit, or all txs have been collected.
+    await Promise.race(this.nodes.map(node => this.collectFastFromNode(request, node, attemptsPerTx, opts)));
+  }
+
+  private async collectFastFromNode(
+    request: FastCollectionRequest,
+    node: TxSource,
+    attemptsPerTx: { txHash: string; attempts: number; found: boolean }[],
+    opts: { deadline: Date },
+  ) {
+    const notFinished = () => this.dateProvider.now() <= +opts.deadline && request.missingTxHashes.size > 0;
+    const maxParallelRequests = this.config.txCollectionFastMaxParallelRequestsPerNode;
+    const activeRequestsToThisNode = new Set<string>(); // Track the txs being actively requested to this node
+    const maxBatchSize = this.config.txCollectionNodeRpcMaxBatchSize;
+
+    const processBatch = async () => {
+      while (notFinished()) {
+        // Pull tx hashes from the attemptsPerTx array, which is sorted by attempts,
+        // so we prioritize txs that have been requested less often.
+        const batch = [];
+        let index = 0;
+        while (batch.length < maxBatchSize) {
+          const txToRequest = attemptsPerTx[index++];
```

Comment:
IIUC this sends the same batch to every node, right? All nodes look at the same list of txs and they all start from index 0 every time and only ensure they don't request the same tx twice.

Woulnd't it be better for different nodes to receives different tx batches? In the best case scenario this will load balance across nodes and all txs would arrive at the same time.

Reply:
> IIUC this sends the same batch to every node

It shouldn't: we're increasing the `attempts` of each tx we picked up (and re-sorting the array) _before_ issuing the request or calling an async function, so when the 2nd node tries to grab the txs it'll see it already modified by the 1st one.